IonDotnet/Internals/Binary/ManagedBinaryWriter.cs

```diff
@@ -116,7 +116,17 @@ public ManagedBinaryWriter(ISymbolTable[] importedTables)
         /// <param name="writeIvm">Whether to write the Ion version marker</param>
         private void StartLocalSymbolTableIfNeeded(bool writeIvm)
         {
-            if (_symbolState != SymbolState.SystemSymbols) return;
+            if (_symbolState == SymbolState.SubsequentLocalSymbols)
```

Comment:
Before you go ahead and update the PR, may I suggest you look at the [java implementation ](https://github.com/amzn/ion-java/blob/7992a822ba85fb538b3dd872f1a86b4d0e80906a/src/software/amazon/ion/impl/bin/IonManagedBinaryWriter.java#L742) which does not need a `SubsequentLocalSymbols` state. Upon a `finish()` (or `flush()`) the state is set back to `SystemSymbols` and the new symbol table (for new symbol) is appended. I think this will cause the Binary version marker to written multiple times (by the call `writeIonVersionMarker()` ) but that shouldn't be a problem if the reader can handle it correctly.

Reply:
I saw the Java implementation and I am puzzled as to how they solve this. It may be that they do not yet support writing this, only reading it, since there appears to be no test for this.
Resetting to SystemSymbols and keeping local symbols would unfortunately break spec, since a version marker is supposed to indicate a new stream (this is how streams can be appended to one another without an understanding of the format).

Reply:
@rickardp You are correct. Writing of local-symbol-table-append syntax is not yet implemented in ion-java through the user-level writers. There is an open [issue](https://github.com/amzn/ion-java/issues/98) to address that (also, [here](https://github.com/amzn/ion-java/pull/111) is the PR that added LST-append read support to ion-java).

It is, however, implemented in ion-c (see [_ion_writer_binary_serialize_symbol_table](https://github.com/amzn/ion-c/blob/master/ionc/ion_writer_binary.c#L1315) for reference). In that implementation, `flush()` causes the writer's symbol table to mark the maximum symbol ID ("`flushed_max_id`") that has been flushed. Subsequent flushes will cause a local symbol table with local-symbol-table-append syntax to be written if `flushed_max_id` is less than the symbol table's total `max_id`.

Your statement about how Ion version markers should be used in relation to `flush()` and `finish()` is also correct. In short, since an Ion version marker always resets the local symbol table, it must always precede the next value written after `finish()`, but not the next value written after `flush()` (so that additional local symbols may be appended).

Reply:
@rickardp It looks like the binary writer would need to be changed to correctly follow the specs and make it possible to test that behavior as well. I imagine the changes could be quite big so if you want to work on it you can update this PR or close this one and create a separate issue to track it. Let me know.

Reply:
@dhhoang One reason to change this PR would be to split it in a reader part and a writer part. This would make sense if one is much more difficult to test. I am happy with either.

As for not following the spec, AFAIK, apart from the obviuous lack of test cases (only one and it’s quite dumb), I believe this PR makes the reader and writer follow that part of the spec. At least I do not expect a huge refactoring of the current solution for the purpose of supporting appending local symbol tables.

If you mean a PR for making sure the binary writer follows spec in its entirety, I would probably argue that would be too big for a single PR, and it would be better to handle each test part of the spec in a different one (where it doesn’t comply).

In any case there is definitely a need for more test cases. I would also like to see tests for Java/JS/C interop where it makes sense.

Reply:
Sorry, now I get what you mean. I based my changes on the assumption that Finish() called Flush(), not the other way around. Yes, this needs to be changed!